accept url query when rank or wolrd_size is specified

chenyangyu1988 opened the pull request:
Summary:
The previously logic will raise exception when there is query in url when rank or world_size is specified
The fix will parse the url and stitch rank and world_size into url.query and regenerate the url.

Test Plan: f161291877

Differential Revision: D19337929

xush6528 commented on torch/distributed/rendezvous.py:
```diff
@@ -54,21 +54,26 @@ def rendezvous(url, rank=-1, world_size=-1, **kwargs):
+            query["world_size"] = world_size
+
+        result = result._replace(
+            query="{}".format("&".join(["{}={}".format(k, v) for k, v in query.items()]))
```
The 2 "query"s here have different types. Can we rename the first one to "query_dict" to avoid confusion?

xush6528 commented on torch/distributed/rendezvous.py:
```diff
@@ -54,21 +54,26 @@ def rendezvous(url, rank=-1, world_size=-1, **kwargs):
+            pair.split("=") for pair in filter(None, result.query.split("&"))
+        )
         assert (
-            "?" not in url
```
For other reviewers:
This is not a good assertion. It effectively disabled adding arguments other than rank and world_size.
This PR is trying to fix it.

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 9e830e18:


* **1/3** failures introduced in this PR
* **2/3** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=9e830e1882461903cfcb0b0e399df42dd8f96df0).

### :snowflake: 2 failures recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4194481) pytorch_linux_xenial_py2_7_9_test (1/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=82951275) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=82951275)) :snowflake:
<details>
<summary>
<code>Jan 10 00:42:31 AssertionError: 1 not less than or equal to 1e-05 :</code>
</summary>

```
Jan 10 00:42:31 ====================================================================== 
Jan 10 00:42:31 FAIL [0.013s]: test_send_recv_with_tag (__main__.TestDistBackend) 
Jan 10 00:42:31 ---------------------------------------------------------------------- 
Jan 10 00:42:31 Traceback (most recent call last): 
Jan 10 00:42:31   File "test_distributed.py", line 2021, in wrapper 
Jan 10 00:42:31     self._join_and_reduce(fn) 
Jan 10 00:42:31   File "test_distributed.py", line 2141, in _join_and_reduce 
Jan 10 00:42:31     self.assertEqual(first_process.exitcode, 0) 
Jan 10 00:42:31   File "/var/lib/jenkins/workspace/test/common_utils.py", line 888, in assertEqual 
Jan 10 00:42:31     super(TestCase, self).assertLessEqual(abs(x - y), prec, message) 
Jan 10 00:42:31 AssertionError: 1 not less than or equal to 1e-05 :  
Jan 10 00:42:31  
Jan 10 00:42:31 ---------------------------------------------------------------------- 
Jan 10 00:42:31 Ran 92 tests in 1.323s 
Jan 10 00:42:31  
Jan 10 00:42:31 FAILED (failures=85, skipped=7) 
Jan 10 00:42:31  
Jan 10 00:42:31 Generating XML reports... 
Jan 10 00:42:31 Traceback (most recent call last): 
Jan 10 00:42:31   File "test/run_test.py", line 456, in <module> 
Jan 10 00:42:31     main() 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4194922) pytorch_linux_xenial_cuda9_cudnn7_py3_multigpu_test (2/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=82954183) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=82954183)) :snowflake:
<details>
<summary>
<code>Jan 10 00:47:21 AssertionError: 1 not less than or equal to 1e-05 :</code>
</summary>

```
Jan 10 00:47:21 ====================================================================== 
Jan 10 00:47:21 FAIL [0.016s]: test_get_rank_size_group (__main__.TestDistBackend) 
Jan 10 00:47:21 ---------------------------------------------------------------------- 
Jan 10 00:47:21 Traceback (most recent call last): 
Jan 10 00:47:21   File "test_distributed.py", line 2021, in wrapper 
Jan 10 00:47:21     self._join_and_reduce(fn) 
Jan 10 00:47:21   File "test_distributed.py", line 2141, in _join_and_reduce 
Jan 10 00:47:21     self.assertEqual(first_process.exitcode, 0) 
Jan 10 00:47:21   File "/var/lib/jenkins/workspace/test/common_utils.py", line 888, in assertEqual 
Jan 10 00:47:21     super(TestCase, self).assertLessEqual(abs(x - y), prec, message) 
Jan 10 00:47:21 AssertionError: 1 not less than or equal to 1e-05 :  
Jan 10 00:47:21  
Jan 10 00:47:21 ====================================================================== 
Jan 10 00:47:21 FAIL [0.012s]: test_reduce_multigpu (__main__.TestDistBackend) 
Jan 10 00:47:21 ---------------------------------------------------------------------- 
Jan 10 00:47:21 Traceback (most recent call last): 
Jan 10 00:47:21   File "test_distributed.py", line 2021, in wrapper 
Jan 10 00:47:21     self._join_and_reduce(fn) 
Jan 10 00:47:21   File "test_distributed.py", line 2129, in _join_and_reduce 
Jan 10 00:47:21     first_process.exitcode == SKIP_IF_SMALL_WORLDSIZE_EXIT_CODE 
Jan 10 00:47:21 AssertionError 
```

</details>

### 1 failure *not* recognized by patterns:
| Job | Step | Status |
| --- | --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_linux_xenial_py3_6_gcc5_4_test</sup> | <sup>Test</sup> | New in PR |


---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

chenyangyu1988 committed:
accept url query when rank or wolrd_size is specified (#32016)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/32016

The previously logic will raise exception when there is query in url when rank or world_size is specified
The fix will parse the url and stitch rank and world_size into url.query and regenerate the url.

Test Plan: f161291877

Differential Revision: D19337929

fbshipit-source-id: 5d8d5c585f71ed6be30aa717c5c4e4c4671e674a

xush6528 commented on torch/distributed/rendezvous.py:
```diff
@@ -131,8 +136,6 @@ def _error(msg):
     def _env_error(var):
         return _error("environment variable %s expected, but not set" % var)
 
-    if not url.startswith("env://"):
```
What's the reason to remove this?

chenyangyu1988 replied:
Remove a unnecessary assert
urllib.parse.urlunparse(parts)
Construct a URL from a tuple as returned by urlparse(). The parts argument can be any six-item iterable. This may result in a slightly different, but equivalent URL, if the URL that was parsed originally had unnecessary delimiters (for example, a ? with an empty query; the RFC states that these are equivalent).
For example:
env:// with rank=1
After urlunparse, the format will be
env?rank=1

chenyangyu1988 replied:
1. Both tcp and file doesn't have such assertion
2. It is what I explained above.

xush6528 reviewed (approved):
Thanks for the fix
